crypto/evp/m_sigver.c

```diff
@@ -36,6 +36,78 @@ static const char *canon_mdname(const char *mdname)
     return mdname;
 }
 
+/*
+ * Handle the case where a key is not exportable from the original provider
+ * pctx: Our EVP_PKEY_CTX
+ * supported_sig: The name of the signature algorithm
+ * tmp_keymgmt: On entry, pointer to the EVP_KEYMGMT that we previously tried to
+ *              export to. On exit, updated to a pointer to the new EVP_KEYMGMT
+ *              (should be the same as the key's)
+ * signature: On exit the EVP_SIGNATURE implementation to use
+ * provkey: On exit updated to point to the provider's key data
+ * Returns:
+ *  1: On success
+ *  0: On soft fail - try legacy
+ * -1: On hard fail
+ */
+int evp_handle_unexportable_key_for_sig(EVP_PKEY_CTX *pctx,
+                                        const char *supported_sig,
+                                        EVP_KEYMGMT **tmp_keymgmt,
+                                        EVP_SIGNATURE **signature,
+                                        void **provkey)
+{
+    const OSSL_PROVIDER *prov;
+
+    /*
+     * Try to get the SIGNATURE operation from the same provider as the
+     * original key (while still respecting the propquery)
+     */
+    prov = EVP_KEYMGMT_get0_provider(pctx->pkey->keymgmt);
+    if (ossl_provider_libctx(prov) != pctx->libctx) {
+        /* We don't support this scenario */
+        return 0;
+    }
+    *signature = evp_signature_fetch_from_prov((OSSL_PROVIDER *)prov,
+                                               supported_sig,
+                                               pctx->propquery);
+    if (*signature == NULL)
+        return 0;
+
+    /*
+     * Now fetch the keymgmt for that provider. This *should* get us the
+     * same keymgmt as used for the key
+     */
+    *tmp_keymgmt = evp_keymgmt_fetch_from_prov((OSSL_PROVIDER *)prov,
+                                               EVP_KEYMGMT_get0_name(*tmp_keymgmt),
+                                               pctx->propquery);
```

Comment:
This still fetches the signature op from default! I think something is not fully right with the `inner_evp_generic_fetch()` modified to take provider argument. It seems to return from the cache without checking for the provider match. If the cache is not filled by a previous call it may seem to work, but not so in my case. Other than that all good -- no calls to EC key generation or keyexch in the provider etc. There is a segfault that is easily fixed (see below).

I don't follow the code well enough to say for sure.